### ISSUE TITLE
Invariance of TypedDict fields makes valid `.update` with another TypedDict fail

### ISSUE THREAD

**Issue by SebbyLaw** [microsoft/pyright#6246]
Introduced from pyright 1.1.332 -> pyright 1.1.333

**Describe the bug**
Invariance on TypedDict fields means valid `.update` operations that previously passed the type checker now report as errors.

**Code or Screenshots**
```python
from typing import TypedDict

class Foo(TypedDict):
    a: int | str
    b: int | str

class Bar(TypedDict):
    a: int
    b: str

foo: Foo = Foo(a=1, b="2")

bar: Bar = Bar(a=2, b="3")

foo.update(bar)  # error here
```
![Code_T9uwr57Av6](https://github.com/microsoft/pyright/assets/44045823/1c84ec54-4030-4c03-9f32-b8d1166cc4a8)

This behavior does not happen with `list.extend()`
```py
l_one: list[str | int] = [1, "2", 3, "4"]
l_two: list[str] = ["a", "b", "c"]

l_one.extend(l_two)  # no error here
```

I believe `.update` on TypedDict should behave the same way as `list.extend()`.

Doing a little dance to not pass a TypedDict passes the type checker:
```py
from typing import TypedDict

class Foo(TypedDict):
    a: int | str
    b: int | str

class Bar(TypedDict):
    a: int
    b: str

foo: Foo = Foo(a=1, b="2")

bar: Bar = Bar(a=2, b="3")

bar_a: int = bar["a"]
bar_b: str = bar["b"]

foo.update({"a": bar_a, "b": bar_b})  # no error here
```

**VS Code extension or command-line**

pyright 1.1.333

**Comment by erictraut** [comment]
I'm not able to repro the problem as reported. Can you please confirm that you're able to repro the problem with the code provided?

Here's your code [sample pasted into pyright playground](https://pyright-play.net/?pyrightVersion=1.1.333&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgCYAiSAxjAFD1UA2AhgM6tQBiYYAFEQjKUaASgBctKFKjMxmFPgA%2BUVjBCTpAIzmolKtfVpM2HAELMQ-EhWoxxGqbPl1pUbfvX1gPOdzBQAXi4eXmYAgEYAGjcAgCIAJliRek0LOXMQQKgM0ID46M04gGYkrx4AOgBXBFJmGGJeVJBkoA). No errors are generated for me.

**Comment by SebbyLaw** [comment]
```
2023-10-27 22:27:39.913 [info] [Info  - 10:27:39 PM] (921887) Pylance language server 2023.10.50 (pyright cfe56bfb) starting
2023-10-27 22:27:39.913 [info] [Info  - 10:27:39 PM] (921887) Server root directory: /home/sebnlaw/.vscode-server/extensions/ms-python.vscode-pylance-2023.10.50/dist
```
My mistake. It seems like a regression between the vscode pylance extension from 2023.10.40 -> 2023.10.50
I'm not sure how to check the versions for these 😅

**Comment by erictraut** [comment]
Ah, I see. Yeah, I'm able to repro the problem with Pylance. If you switch to the prerelease version of Pylance (v2023.10.53), the problem will go away.

I typically release pyright once a week (on Tues evening). Pylance likewise releases once a week (on Wed morning). We coordinate the timing to reduce any differences between pyright and pylance. 

Digging into this a bit deeper, here's what appears to have happened in this case. I released pyright version 1.1.332 on the evening of Tues Oct 17. On the morning of Wed Oct 18, I committed another change (a bug fix), and this change introduced a regression that explains what you're seeing. I found and fixed the regression a day later (which explains why the problem doesn't occur with pyright 1.1.333), but that was too late. The Pylance team had already snapshotted the pyright repo later on Wed morning. They picked up this regression in the pylance prerelease build v2023.10.31. This build was automatically promoted to the released version v2023.10.40 the following week (on Wed Oct 24). Version v2023.10.50 was a hot-fix, so it was based on the same code as v2023.10.40.

@debonte & @rchiodo, maybe we should consider changing the Pylance build process to use the last-tagged version of the pyright repo (i.e. the commit that corresponds to the last official release of pyright) rather than the tip of tree? That would eliminate issues like this and increase compatibility between pyright and pylance releases. Thoughts?

An extension to this idea... if we more carefully align pylance and pyright builds, would it make sense to add a command to pylance so users can easily determine the pyright version used in that release of pylance? Or alternatively, pylance could output the pyright version in its logs.

**Comment by rchiodo** [comment]
> Or alternatively, pylance could output the pyright version in its logs.

We output the commit id at the moment. Seems like it would make more sense to output the version:

```
2023-10-27 07:58:12.892 [info] [Info  - 07:58:12] (38682) Pylance language server 2023.9.10 (pyright c70adefc) starting
```
